### ISSUE TITLE
CA2213 triggers on a compiler-generated backing field

### ISSUE THREAD

**Issue by bkoelman** [dotnet/roslyn#4077]
The built-in analyzer for CA2213 in VS2015 RTM reports a diagnostic on a compiler-generated backing field, while its declared property is correctly disposed of. So the diagnostic should not appear in this case.

![ca2213-backing-field](https://cloud.githubusercontent.com/assets/10324372/8851264/b0cd629a-314e-11e5-9f37-9a6f82ced475.png)

**Comment by breyed** [comment]
Even better, CA2213  should check that the property associated with the backing field has `Dispose` or `Close` called on its object.

**Comment by srivatsn** [comment]
Can you try if installing the [FxCopAnalyzers package](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers/1.0.1) to see if that fixes the issue for you? The FxCopAnalyzers package reimplements some of the CA rules using roslyn and that's where we are making changes.

**Comment by jahmai** [comment]
It really doesn't matter whether this works or not because: 

- It installs 4 (!) nuget dependencies per project (!) 
- It depends on assemblies that aren't portable, so doesn't work on WinPhone or Xamarin. 

We've just disabled CA2213 until it is addressed in a Visual Studio update or something. 

----- Original Message -----

**Comment by bkoelman** [comment]
@srivatsn Thanks for addressing this. After installing v1.0.1 the warning is no longer raised. However, that package is causing another issue on my code (see https://github.com/dotnet/roslyn-analyzers/issues/281).

**Comment by jmarolf** [comment]
@jahmai The analyzer dlls are installed per project, but they are not packaged with your application, is there are limit in your organization regarding the number of nuget packages you can use?.  Also, could you elaborate on what you mean by: " It depends on assemblies that aren't portable, so doesn't work on WinPhone or Xamarin."?  Windows Phone and Xamarin projects support analyzers, or are you referring to an environment other then Visual Studio?

**Comment by breyed** [comment]
Thinking of Code Analysis as an extension of the compiler (deep warnings), it's seems odd to use NuGet to distribute CA updates. A Visual Studio extension seems a better fit. For example, F# is distributed as an extension. Using NuGet packages allows you to have different versions of CA per project, which means you need to keep versions up to date for every project and solution.

**Comment by jmarolf** [comment]
@breyed The thinking here was that we didn't want people's machine configuration to change their code analysis results.  With nuget, your team can decide on a version of the FXCop packages and checkin a packages.config file into source control.  That way all builds should be using the same tools and get the same results.

**Comment by breyed** [comment]
@jmarolf Good point. It's a trade-off. You can have per-project or per-user updates be automatic, but not both. For infrequently updated tools, like the C# compiler, per-user is great: install VS2015 once, and you are set for all your projects. If CA is updated frequently, then per-project is nice: one dev on the team updates it and checks in a `packages.config` update, and the rest of the team gets the update.

The second scenario is not so clean, however. Realistically, a dev likely has multiple projects, at least one of which is manages himself. So he's doing the "per-user" update anyway, since he only "inherits" updates for a subset of all this projects.

The per-user approach doesn't guarantee everyone is on the same version. However, a simple policy of, "when Visual Studio notifies you of an CA extension update, install it", would keep everyone rather well in sync. These are just warnings, so perfect compatibility is not essential.

**Comment by jmarolf** [comment]
"These are just warnings, so perfect compatibility is not essential." As an aside, many organizations build with warnings-as-errors, so even introducing warnings can be very destructive for people.

I think this is worthy of its own bug (or set of bugs).  Ideally the anyone should be able to choose somewhere on the continuum between Dev-local and project-local and things just work.  Here's what I'm hearing:

1. Updating or just managing analyzers at a solution level is a pain
2. There is no clean way to manage "dev local" settings and "project settings" when it comes to analyzers

Is there anything else?

**Comment by breyed** [comment]
Warnings-as-errors is great for keeping the build clean, although it is fair to temporarily relax the policy if a tooling update throws a number of new warnings at you. You could roll back the tool (or not update it) to suppress the warnings, but if the warnings are suppressed, you've already lost much of the value of the warnings-as-errors policy.

Thinking about warnings-as-errors reminded me of the automated build server, which gives me a thought to your "anything else" question: You have to think about update procedures for build servers. It's one thing to patch them for major updates (VS2015.1, .2, etc.), but it would be more friction to have to patch them whenever an extension update is published.

Update friction doesn't apply to hosted build servers; one would hope a they would always be up to date. However, there's a new problem: everyone always gets the edge versions. The "forced upgrade" nature of a hosted build server makes me think that "per-user" CA updates should be infrequent and take the slower, more thoroughly betaed approach that the compiler takes.

This leads me to see two models:

1. "In-the-box" updates that ship with the VS updates. Well tested. Very stable. Updated only once per quarter or so, along with VS.

2. NuGet updates that come out whenever. When applied, these override the "in-the-box" versions. These can be more bleeding edge.

Hopefully, the delta between the two models is small enough that most people can blissfully just go with model 1 and not worry about the extra configuration involved in 2. Model 2 is great for enthusiasts and for beta testing. God forbid it should become an excuse for slack in the testing and quality of model 1.

The one model that doesn't seem to make sense (having thought it through more) is a VS extension. It's an anti-pattern because of the instability it would cause on hosted build servers.

**Comment by jmarolf** [comment]
alright, I think this is getting off-topic for this issue.  I'll file the bugs and tag you @breyed so we can continue this discussion.

**Comment by lgolding** [comment]
The original issue is now [resolved](https://github.com/dotnet/roslyn/issues/4077#issuecomment-136318402).

Per @jmarolf's [comment](https://github.com/dotnet/roslyn/issues/4077#issuecomment-136496757), closing this issue.
